sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnVector.java

```diff
@@ -136,7 +136,7 @@ public int getInt(int rowId) {
   public long getLong(int rowId) {
     int index = getRowIndex(rowId);
     if (isTimestamp) {
-      return timestampData.time[index] * 1000 + timestampData.nanos[index] / 1000;
+      return timestampData.time[index] * 1000 + timestampData.nanos[index] / 1000 % 1000;
```

Comment:
In Apache ORC 1.4.4, ORC-306 fixes this according to the [original definition](https://github.com/apache/hive/blob/master/storage-api/src/java/org/apache/hadoop/hive/ql/exec/vector/TimestampColumnVector.java#L45-L46).

Reply:
Add a test case?

Reply:
Do you know when this issue was introduced in ORC?

Reply:
The change is on [TreeReaderFactory.java](https://github.com/apache/orc/commit/6c4865ad9acb75c35d97206f31b4dd9e0a3a7cb4#diff-dcf15a871eb200f0fceaa924e14a01d4R980). From Apache ORC project, the prior code is ORC-1 which is importing code from Hive two years ago.

Effectively, the writer side is the same. Only, reader side is changed.

Reply:
`OrcHadoopFsRelationSuite` covers this changes via end-to-end write and read test cases.

Reply:
Based on my understanding, ORC-306 changes the query result, right?

Reply:
ORC-306 changes the content of exposed ORC column vectors in reader side. The interpretation is Spark's logic as we see in this PR.

Reply:
Are you saying no external impact of ORC-306?

Reply:
No, what I mean is, with ORC-306 and this fix, there is no external impact outside Spark. More specifically, outside `OrcColumnVector`/`OrcColumnarBatchReader`. In other words, ORC 1.4.4 cannot be used with Apache Spark without this patch.

Java `Timestamp.getTime` and Timestamp.getNano` has an overlap by definition. Previously, ORC didn't stick to the definition.